knimin/lib/data_access.py

```diff
@@ -2759,6 +2767,155 @@ def delete_samples(self, ids):
                          RETURNING sample_id"""
                 TRN.add(sql, [sample_id])
 
+    def _study_sample_are_associated(self, study_id, sample_id):
+        """Checks whether a study and a sample are associated
+
+        Parameters
+        ----------
+        study_id : str
+            ID of the study to check
+        sample_id : str
+            ID of the sample to check
+
+        Returns
+        ------
+        bool
+            Whether they are associated
+        """
+        sql = """SELECT EXISTS (SELECT 1 FROM pm.study_sample
+                                WHERE study_id = %s
+                                AND sample_id = %s)"""
+        return self._con.execute_fetchone(sql, [study_id, sample_id])[0]
+
+    def associate_samples_with_study(self, study_id, sample_ids):
+        """Creates associations between one or more samples and one study
+
+        Parameters
+        ----------
+        study_id : int
+            ID of the study
+        sample_ids : list of str
+            IDs of the samples
+
+        Raises
+        ------
+        ValueError
+            If the study ID does not exist
+            If the study and the sample are already associated
+        """
+        self._study_exists(study_id)
+        with TRN:
+            for sample_id in sample_ids:
+                if not self._sample_exists(sample_id):
+                    raise ValueError('Sample ID %s does not exist.'
+                                     % sample_id)
+                if self._study_sample_are_associated(study_id, sample_id):
+                    raise ValueError('Study ID %s and sample ID %s are already'
+                                     ' associated.' % (study_id, sample_id))
+                sql = """INSERT INTO pm.study_sample (study_id, sample_id)
+                         VALUES (%s, %s)"""
+                TRN.add(sql, [study_id, sample_id])
+
+    def dissociate_samples_from_study(self, study_id, sample_ids):
```

Comment:
What happens if we end up with orphan samples? I know I missed a couple of PRs, but in my opinion we work in a "study" basis rather than in a "sample" basis. I.e. we are not going to process samples if we don't know to which study they belong to.